src/storage/pooled_storage_manager.h

```diff
@@ -25,372 +25,418 @@
-    profiler::GpuDeviceStorageProfiler::Get()->OnFree(handle);
-  }
-
- private:
   // log2 of maximum page size. 16GB
   const size_t LOG2_MAX_MEM = 34;
-  // address width in bits
-  static const int addr_width = sizeof(size_t) * 8;
-  // used memory
-  size_t used_memory_ = 0;
-  // page size
-  size_t page_size_;
   // log2 of memory size before switching to exponential mode to linear mode
-  size_t cut_off_;
-  // percentage of reserved memory
-  int reserve_;
-  // context used by this Storage Manager
-  const Context initial_context_;
-  // memory pool
-  std::vector<std::vector<void*>> memory_pool_;
-  DISALLOW_COPY_AND_ASSIGN(GPUPooledRoundedStorageManager);
-};  // class GPUPooledRoundedStorageManager
-
-void GPUPooledRoundedStorageManager::Alloc(Storage::Handle* handle) {
-  // Set dptr to nullptr when handle size is 0.
-  if (handle->size == 0) {
-    handle->dptr = nullptr;
-    return;
+  size_t cut_off_ = 0;
+};  // class RoundPower2
+
+
+/*!
+ * \brief Unordered map based storage container.
+ *  The pointers to the portions of same rounded sizes memory
+ *  allocated on CPU/GPU, are stored in separate vectors.
+ *  These sizes are used as keys for accessing the vectors,
+ *  which are the elements stored in an unordered map.
+ */
+class UnorderedMapContainer {
+ protected:
+  inline void InitContainer(const RoundHelper *p)   {}
+  inline void InsertInCache(size_t key, void *dptr) { memory_pool_[key].push_back(dptr); }
+
+  inline std::vector<void*> *GetMemStorage(size_t key) {
+    auto&& reuse_it = memory_pool_.find(key);
+    return reuse_it != memory_pool_.end() && reuse_it->second.size()? &reuse_it->second : nullptr;
   }
 
-  std::lock_guard<std::mutex> lock(Storage::Get()->GetMutex(Context::kGPU));
-  int bucket = get_bucket(handle->size);
-  size_t size = get_size(bucket);
-  auto&& reuse_pool = memory_pool_[bucket];
-  if (reuse_pool.size() == 0) {
-    mxnet::common::cuda::DeviceStore device_store(handle->ctx.real_dev_id(), true);
-    size_t free, total;
-    cudaMemGetInfo(&free, &total);
-    if (free <= total * reserve_ / 100 || size > free - total * reserve_ / 100)
-      ReleaseAll();
-
-    void* ret = nullptr;
-    cudaError_t e = cudaMalloc(&ret, size);
-    if (e != cudaSuccess) {
-      if (e == cudaErrorMemoryAllocation) {
-        ReleaseAll();
-        e = cudaMalloc(&ret, size);
-        if (e != cudaSuccess && e != cudaErrorCudartUnloading) {
-          LOG(FATAL) << "cudaMalloc retry failed: " << cudaGetErrorString(e);
-        }
-      } else if (e != cudaErrorCudartUnloading) {
-        LOG(FATAL) << "cudaMalloc failed: " << cudaGetErrorString(e);
+  size_t ReleaseAllNoLock(const ContextHelper *contextHelper, const RoundHelper * /*rndHelper*/) {
+    SET_GPU_PROFILER(profilerGPU, contextHelper);
+    size_t released_memory = 0;
+    for (auto&& i : memory_pool_) {
+      for (auto&& j : i.second) {
+        contextHelper->Free(j);
+        GPU_PROFILER_ON_FREE(profilerGPU, j);
       }
```

Comment:
Actually, new code contains exactly the same retry (see `retry in case of fragmentation` comment)
```cpp
    roundSize = BucketingStrategy::RoundAllocSizeForBucket(bucket_id);
    if (!MemoryIsAvalable(roundSize))
      ReleaseAllNoLock(false);

    void *ret = nullptr;
    auto e = contextHelper_->Malloc(&ret, roundSize);
    if (e) {
      // retry in case of fragmentation
      ReleaseAllNoLock(false);
      e = contextHelper_->Malloc(&ret, roundSize);
      if (e) {
        const std::string err(
#if MXNET_USE_CUDA
        dev_type_ == Context::kGPU?
           cudaGetErrorString(static_cast<cudaError_t>(e)) :
#endif
           std::strerror(errno));

        LOG(FATAL) << "Memory allocation failed " << err;
      }
    }

    UNSET_DEVICE(device_store);
    used_memory_ += roundSize;
```

As you could see, in that part I am using `ReleaseAllNoLock(false)` twice: 

- when the amount of free memory is below some threshold
- after unsuccessful first attempt to allocate memory.
After each of these two calls, we try to allocate memory by `contextHelper_->Malloc(&ret, roundSize);`

Previously, we had exactly the same:
```cpp
    size_t free, total;
    cudaMemGetInfo(&free, &total);
    if (free <= total * reserve_ / 100 || size > free - total * reserve_ / 100)
      ReleaseAll();

    void* ret = nullptr;
    cudaError_t e = cudaMalloc(&ret, size);
    if (e != cudaSuccess) {
      if (e == cudaErrorMemoryAllocation) {
        ReleaseAll();
        e = cudaMalloc(&ret, size);
        if (e != cudaSuccess && e != cudaErrorCudartUnloading) {
          LOG(FATAL) << "cudaMalloc retry failed: " << cudaGetErrorString(e);
        }
      } else if (e != cudaErrorCudartUnloading) {
        LOG(FATAL) << "cudaMalloc failed: " << cudaGetErrorString(e);
      }
    }
    used_memory_ += size;
```

Reply:
It wasn't in your PR https://github.com/andrei5055/incubator-mxnet/blob/97dfc823d8b89751ff8be459a0357195e383a23d/src/storage/pooled_storage_manager.h#L178-L189

I added it in https://github.com/apache/incubator-mxnet/pull/18987/files

Reply:
Oh, sorry, I didn't notice that it was your change.

In fact, my initial @NVidia implementation had these or very similar lines (and it's why I thought these lines were from my PR). But @ptrendx convinced me that they are useless. Also, I've done some experiments trying to find real-life cases where these lines actually work, and couldn't find any such cases. So I decided not to include them in my PR.